pkg/cluster/agent_customization.go

```diff
@@ -0,0 +1,86 @@
+package cluster
+
+import (
+	"encoding/json"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// GetClusterAgentTolerations returns additional tolerations for the cluster agent if they have been user defined. If
+// not, nil is returned.
+func GetClusterAgentTolerations(cluster *v3.Cluster) []corev1.Toleration {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.AppendTolerations
+	}
+
+	return nil
+}
+
+// GetClusterAgentAffinity returns node affinity for the cluster agent if it has been user defined. If not, then the
+// default affinity is returned.
+func GetClusterAgentAffinity(cluster *v3.Cluster) *corev1.Affinity {
+	if cluster.Spec.ClusterAgentDeploymentCustomization != nil &&
+		cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity != nil {
+		return cluster.Spec.ClusterAgentDeploymentCustomization.OverrideAffinity
+	}
+
+	return unmarshalAffinity(settings.ClusterAgentDefaultAffinity.Get())
+}
```

Comment:
@jakefhyde , tbh I thought of the same while reviewing it but deemed it "good enough" as I could argue that falling back to default affinity in case of not properly formatted updated setting value could actually be more harmful than falling back to nil affinity as it's less obvious that something's not right.
The proper way to handle this would be some mechanism to validate setting values on startup but that's asking too much as part of this particular enhancement.

So as soon as we log these in a helpful fashion it's debatable what is the best approach - I'm not sure how good are errors returned by `json.Unmarshal` (now that I saw that I see a minor issue).